src/V3Hasher.cpp

```diff
@@ -29,40 +31,290 @@ class HasherVisitor final : public AstNVisitor {
-            }
+    inline void walk(AstNode* nodep) {  //
+        if (VL_LIKELY(nodep)) iterate(nodep);
+    }
+
+    V3Hash hashNode(AstNode* nodep, bool hashDType, bool hashChildren, std::function<void()>&& f) {
+        if (m_cacheInUser4 && nodep->user4()) {
+            return V3Hash(nodep->user4());
+        } else {
+            VL_RESTORER(m_hash);
+            // Reset accumulator
+            m_hash = V3Hash(nodep->type());  // Node type
+            f();  // Node specific hash
+            if (hashDType && nodep != nodep->dtypep()) walk(nodep->dtypep());  // Node dtype
+            if (hashChildren) iterateChildrenConst(nodep);  // Children
+            if (m_cacheInUser4) { nodep->user4(m_hash.value()); }
+            return m_hash;
         }
-        // Update what will become the above node's hash
-        m_lowerHash += m_cacheInUser4 ? V3Hash(nodep->user4()) : thisHash;
     }
 
+    // VISITORS
+
+    // Each visitor below contributes to the hash any node specific content
+    // that is not dependent on either of the following, as these are
+    // included by default by hashNode:
+    // - Node type (as given by AstNode::type())
+    // - Node dtype (unless !hashDType)
+    // - child nodes (unless !hashChildren)
+    //
+    // The hash must be stable, which means in particular it cannot rely on
+    // pointer values, or any other value that might differ between separate
+    // invocations of Verilator over the same design.
+    //
+    // Note there is a circularity problem where some child nodes can back
+    // to their ancestral nodes via member pointers, which can lead to an
+    // infinite traversal. To break this, nodes that are subject to such
+    // referencing and represent code which can reasonably be assumed not to
+    // be equivalent to any other code, are hashed either by name (e.g.:
+    // AstNodeModule), or by unique identifier (e.g.: AstNodeUOrStructDType).
+
+#define VISIT(kind, hashDType, hashChildren, body) \
+    virtual void visit(kind* nodep) override { \
+        const V3Hash hash = hashNode(nodep, hashDType, hashChildren, [=]() body); \
+        m_hash += hash; \
+    }
+
+    //------------------------------------------------------------
+    // AstNode - Warns to help find missing cases
+
+    VISIT(AstNode, true, true, {  //
+        UINFO(4, "WARNING! Hashing node as AstNode: " << nodep);
+    })
+
+    //------------------------------------------------------------
+    // AstNodeDType
+    VISIT(AstNodeArrayDType, false, true, {
+        walk(nodep->virtRefDTypep());
+        m_hash += nodep->left();
+        m_hash += nodep->right();
+    })
+    VISIT(AstNodeUOrStructDType, false, false, {  //
+        m_hash += nodep->uniqueNum();
+    })
+    VISIT(AstParamTypeDType, true, true, {
+        m_hash += nodep->name();
+        m_hash += nodep->varType();
+    })
+    VISIT(AstMemberDType, false, true, {  //
+        m_hash += nodep->name();
+    })
```

Comment:
I'm wondering if this is harder to maintain versus for moving most of these back into the ast classes and calling a common virtual function from this visitor's default AstNode visitor? Having most in the classes seems easier to maintain (or at least harder to forget to add it.)

Reply:
> I'm wondering if this is harder to maintain versus for moving most of these back into the ast classes and calling a common virtual function from this visitor's default AstNode visitor?

Actually on this one I couldn't disagree more. I think it's great for maintenance to have all code related to a single functionality (in this case tree hashing) be in a singel file, all together. Sadly C++ evonving a the peak of OOP, there are few nice mechanisms in the language to do this, and you end up going the way of splitting your functionality into onelines spread over hundreds of classes. I'm actually a bit surprised you would advocate for "just glom these into the already 3000 and 9000+ line long global headers and spread them out semi regularly throughout the file, that will make it more maintainable" ;). The V3Ast*.h headers are already borderline impenetrable, so getting things out of there would be an improvement IMO.

As a curiosity, this is the AST description for Alogic for comparison:
https://github.com/ArgonDesign/alogic/blob/914059911e73426fc2c3ad8cc117a221ffb7de11/src/main/scala/com/argondesign/alogic/ast/Trees.scala#L35-L369

And here are for example two predicates dealing with whether a procedural statement can terminate control flow, all nicely separate:
https://github.com/ArgonDesign/alogic/blob/914059911e73426fc2c3ad8cc117a221ffb7de11/src/main/scala/com/argondesign/alogic/ast/StmtOps.scala#L14-L46

Maybe my mind is poisoned, but I wish C++ had `match` :)

> or at least harder to forget to add it.

Actually, I think you ('you' as in anyone but Wilson) are almost guaranteed to forget because the base classes have implementations of the virtual function so the compiler doesn't help you. With this at least the base case `visit(AstNode*)` still nags in a more direct way, compared to what `AstNode::sameHash` used to do.

Reply:
Actually forgot to mention the technical reason I extracted these, apart from the philosophical ones. The nodes need to be able to invoke iterate on non-child members, which means the `sameHash` would have to be passed the `HasherVisitor`, which I think would make the headers painfully circular.